src/engine.rs

```diff
@@ -191,7 +224,17 @@ where
             debug_assert_eq!(writer.perf_context_diff.write_wait_duration, Duration::ZERO);
             perf_context += &writer.perf_context_diff;
             set_perf_context(perf_context);
-            writer.finish()?
+            // Retry if `writer.finish()` returns a special err, remarking there still
+            // exists free space for this `LogBatch`.
+            let ret = writer.finish();
+            if let Err(Error::Other(_)) = ret {
+                log_batch.reset_to_encoded_state();
+                // Here, we will retry this LogBatch `append` by appending this writer
+                // to the next write group, and the current write leader will not hang
+                // on this write and will return timely.
+                return self.write(log_batch, sync);
```

Comment:
This will probably enter infinite recursive calls if the disk is still full after rotate.

Reply:
Just like your concerns in the following comments `Two things here ...`, I still wonder whether there exists other elegant ways to handle it.

Reply:
This issue is not addressed.